### PR TITLE
[AGW] Add Fua Service Restriction proposal

### LINKED ISSUE #2719: [SessionD] Add FUA-Restrict Support
This project tracks the FUA-Restrict progress separately: https://github.com/magma/magma/projects/7#card-45998136

### PR DESCRIPTION
**Pull request by ymasmoudi**
## Summary

A proposal for Service Restriction feature support.

## Test Plan

NaN

### CODE REVIEW

File: docs/proposals/fua-restrict-feature/fua-restrict-feature.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+**Session Proxy**
+
+We will limit this implementation to Filter-Id AVP only as It covers the desired
+use-cases. The AVP will reference a PolicyRule Identifier.
+
+```
+  [Final-Unit-Indication]
+    {Final-Unit-Action}
+    *[Filter-Id]
```
Comment (themarwhal) [inline docs/proposals/fua-restrict-feature/fua-restrict-feature.md:72]: 'Filter-Id' will correspond to a PolicyRule's ID field, right?

File: docs/proposals/fua-restrict-feature/fua-restrict-feature.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+In addition, we will need to store the list of rules ID to be used when restriction
+is enabled. A method get\_restricted\_rule\_ids will be added to retrive this list
+prior to sending rules activation to pipelined.
+
+We will re-use the same pipelined client method add\_gy\_final\_action\_flow as It
+provides support for this case as well.
+
+Static Rules will be sent using their original priority and will be applied as gy
+final action flow.
+
```
Comment (themarwhal) [inline docs/proposals/fua-restrict-feature/fua-restrict-feature.md:99]: Meaning we will not do any priority manipulation in SessionD for static rules? We can leave the static rule definitions as is. And we are not attempting to resolve the `Restriction-Filter-Rule` (dynamic rule) in this change? (Sorry, if this is already specified in the doc. Just trying to make sure I'm comprehending it correctly :) )

File: docs/proposals/fua-restrict-feature/fua-restrict-feature.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+The service state SERVICE\_RESTRICTED will need to be supported as well.
+
+
+**Pipelined**
+
+GYController needs to be updated to support restriction FUA. This translates to,
+1/ activate all the flows as they are received from SessionD.
+2/ Support QoS in GY APP (Include a Qos Manager)
+3/ Updating statistics back to SessionD
```
Comment (themarwhal) [inline docs/proposals/fua-restrict-feature/fua-restrict-feature.md:108]: I think if we are sending up stats for Gy rules, we will need some changes in SessionD as well to apply the Gy stats in. (Shouldn't be a big change, but we don't look at Gy rules when we currently take in enforcement stats)

**Review by themarwhal — APPROVED** [review]

File: docs/proposals/fua-restrict-feature/fua-restrict-feature.md (hunk 1 above)
Comment (ymasmoudi) [reply docs/proposals/fua-restrict-feature/fua-restrict-feature.md:72]: Yes, that's correct

File: docs/proposals/fua-restrict-feature/fua-restrict-feature.md (hunk 3 above)
Comment (ymasmoudi) [reply docs/proposals/fua-restrict-feature/fua-restrict-feature.md:108]: Yes, good point. I will add this to the list

File: docs/proposals/fua-restrict-feature/fua-restrict-feature.md (hunk 2 above)
Comment (ymasmoudi) [reply docs/proposals/fua-restrict-feature/fua-restrict-feature.md:99]: Yes, exactly. The first implementation will support Filter-Id only as It covers all our requirements. Thus, we will use the static rules with their predefined priorities without changes.

**Review by themarwhal — APPROVED** [review]

**Pull request merged**